pkg/solana/logpoller/loader.go

```diff
@@ -0,0 +1,265 @@
+package logpoller
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+type Parser interface {
+	// ProcessEvent should take a ProgramEvent and parse it based on log signature
+	// and expected encoding. Only return errors that cannot be handled and
+	// should exit further transaction processing on the running thread.
+	//
+	// ProcessEvent should be thread safe.
+	ProcessEvent(ProgramEvent) error
+}
+
+type RPCClient interface {
+	GetLatestBlockhash(ctx context.Context, commitment rpc.CommitmentType) (out *rpc.GetLatestBlockhashResult, err error)
+	GetBlocks(ctx context.Context, startSlot uint64, endSlot *uint64, commitment rpc.CommitmentType) (out rpc.BlocksResult, err error)
+	GetBlockWithOpts(context.Context, uint64, *rpc.GetBlockOpts) (*rpc.GetBlockResult, error)
+	GetSignaturesForAddressWithOpts(context.Context, solana.PublicKey, *rpc.GetSignaturesForAddressOpts) ([]*rpc.TransactionSignature, error)
+	GetTransaction(context.Context, solana.Signature, *rpc.GetTransactionOpts) (*rpc.GetTransactionResult, error)
+}
+
+const (
+	DefaultNextSlotPollingInterval = 1_000 * time.Millisecond
+)
+
+type EncodedLogCollector struct {
+	// service state management
+	services.Service
+	engine *services.Engine
+
+	// dependencies and configuration
+	client       RPCClient
+	parser       Parser
+	lggr         logger.Logger
+	rpcTimeLimit time.Duration
+
+	// internal state
+	chSlot  chan uint64
+	chBlock chan uint64
+	chJobs  chan Job
+	workers *WorkerGroup
+
+	mu                sync.RWMutex
+	loadingBlocks     atomic.Bool
+	highestSlot       uint64
+	highestSlotLoaded uint64
+	lastSentSlot      atomic.Uint64
+}
+
+func NewEncodedLogCollector(
+	client RPCClient,
+	parser Parser,
+	lggr logger.Logger,
+) *EncodedLogCollector {
+	c := &EncodedLogCollector{
+		client:       client,
+		parser:       parser,
+		chSlot:       make(chan uint64, 1),
+		chBlock:      make(chan uint64, 1),
+		chJobs:       make(chan Job, 1),
+		lggr:         lggr,
+		rpcTimeLimit: 1 * time.Second,
+	}
+
+	c.Service, c.engine = services.Config{
+		Name: "EncodedLogCollector",
+		NewSubServices: func(lggr logger.Logger) []services.Service {
+			c.workers = NewWorkerGroup(DefaultWorkerCount, lggr)
+
+			return []services.Service{c.workers}
+		},
+		Start: c.start,
+		Close: c.close,
+	}.NewServiceEngine(lggr)
+
+	return c
+}
+
+func (c *EncodedLogCollector) BackfillForAddress(ctx context.Context, address string, fromSlot uint64) error {
```

Comment:
I don't see anything calling this, have you tried testing it yet?

Reply:
I don't have a direct test for it yet. It uses roughly the same internal mechanisms, but I still need a smoke test at minimum.